src/main/kotlin/org/rust/ide/template/postfix/MatchPostfixTemplate.kt

```diff
@@ -32,31 +39,83 @@ class MatchPostfixTemplate(provider: RsPostfixTemplateProvider) :
-        }
-        fix.invoke(project, matchExpr.containingFile, matchExpr, matchExpr)
+        processor.normalizeMatch(matchExpr)
 
         PsiDocumentManager.getInstance(project).doPostponedOperationsAndUnblockDocument(editor.document)
 
         val matchBody = matchExpr.matchBody ?: return
-        val wildPatterns = matchBody.descendantsOfType<RsPatWild>()
+        val toBeReplaced = processor.getElementsToReplace(matchBody)
 
-        if (wildPatterns.isEmpty()) {
+        if (toBeReplaced.isEmpty()) {
             val arm = matchBody.matchArmList.firstOrNull() ?: return
             val blockExpr = arm.expr as? RsBlockExpr ?: return
             editor.caretModel.moveToOffset(blockExpr.block.lbrace.textOffset + 1)
         } else {
-            editor.buildAndRunTemplate(matchBody, wildPatterns.map { it.createSmartPointer() })
+            editor.buildAndRunTemplate(matchBody, toBeReplaced.map { it.createSmartPointer() })
         }
     }
 }
+
+private fun getMatchProcessor(ty: Ty, context: RsElement): MatchProcessor {
+    return when {
+        ty is TyAdt && ty.item == context.knownItems.String -> StringMatchProcessor
+        ty is TyReference && ty.referenced is TyStr -> StringLikeMatchProcessor()
+        else -> GenericMatchProcessor
+    }
+}
+
+private abstract class MatchProcessor {
+    abstract fun createMatch(factory: RsPsiFactory, expression: RsExpr): RsMatchExpr
+    open fun normalizeMatch(matchExpr: RsMatchExpr) {}
+
+    open fun getElementsToReplace(matchBody: RsMatchBody): List<RsElement> = emptyList()
+}
+
+private object GenericMatchProcessor : MatchProcessor() {
+    override fun createMatch(factory: RsPsiFactory, expression: RsExpr): RsMatchExpr {
+        val exprText = if (expression is RsStructLiteral) "(${expression.text})" else expression.text
+        return factory.createExpression("match $exprText {}") as RsMatchExpr
+    }
+
+    override fun normalizeMatch(matchExpr: RsMatchExpr) {
+        val patterns = matchExpr.checkExhaustive().orEmpty()
+        val fix = if (patterns.isEmpty()) {
+            AddWildcardArmFix(matchExpr)
+        } else {
+            AddRemainingArmsFix(matchExpr, patterns)
+        }
+        fix.invoke(matchExpr.project, matchExpr.containingFile, matchExpr, matchExpr)
+    }
+
+    override fun getElementsToReplace(matchBody: RsMatchBody): List<RsElement> =
+        matchBody.descendantsOfType<RsPatWild>().toList()
+}
+
+private open class StringLikeMatchProcessor : MatchProcessor() {
+    open fun expressionToText(expression: RsExpr): String = expression.text
+
+    override fun createMatch(factory: RsPsiFactory, expression: RsExpr): RsMatchExpr {
+        val exprText = expressionToText(expression)
+        return factory.createExpression("match $exprText {\n\"\" => {} }") as RsMatchExpr
+    }
+
+    override fun normalizeMatch(matchExpr: RsMatchExpr) {
+        AddWildcardArmFix(matchExpr).invoke(matchExpr.project, matchExpr.containingFile, matchExpr, matchExpr)
```

Comment:
Why not just produce a `match` with an additional wildcard arm in `createMatch`?

Reply:
Hmm, I was a bit worried about formatting, but it works fine. I changed it to generate `_` directly.